clientv3/ordering/kv.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2017 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ordering
+
+import (
+	"context"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/etcdserver/api/v3rpc/rpctypes"
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+)
+
+type kvOrderingCache struct {
+	clientv3.KV
+	switchEndpoint SwitchEndpointClosure
+	prevRev        int64
+}
+
+// NewKVOrderingCache wraps a KV instance so that all
+// get and txn responses are guaranteed to have revisions
+// that are equal to or greater than the previous known revision
+// by cache the previous known revision.
+func NewKV(kv clientv3.KV, f SwitchEndpointClosure) *kvOrderingCache {
+	return &kvOrderingCache{kv, f, 0}
+}
+
+func (kv *kvOrderingCache) Get(ctx context.Context, key string, opts ...clientv3.OpOption) (*clientv3.GetResponse, error) {
+	if len(key) == 0 {
+		return nil, rpctypes.ErrEmptyKey
+	}
+	r, err := kv.KV.Do(ctx, clientv3.OpGet(key, opts...))
```

Comment:
get the previous revision before submitting the RPC; otherwise, this breaks with concurrent accesses. Example:req1 has prevRev=1, req2 has prevRev=1, req2 returns and sets prevRev=2, req1 returns and uses prevRev=2 when it should be comparing with prevRev=1